File: src/main/java/com/gridgame/common/model/Position.java
package com.gridgame.common.model;

import com.gridgame.common.Constants;
import java.util.Objects;

/**
 * Immutable position on the game grid.
 * Coordinates are validated to be within bounds [0, GRID_SIZE).
 */
public final class Position {
    private final int x;
    private final int y;

    public Position(int x, int y) {
        if (x < 0 || x >= Constants.GRID_SIZE) {
            throw new IllegalArgumentException(
                String.format("X coordinate %d out of bounds [0, %d)", x, Constants.GRID_SIZE));
        }
        if (y < 0 || y >= Constants.GRID_SIZE) {
            throw new IllegalArgumentException(
                String.format("Y coordinate %d out of bounds [0, %d)", y, Constants.GRID_SIZE));
        }
        this.x = x;
        this.y = y;
    }

    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (!(obj instanceof Position)) return false;
        Position other = (Position) obj;
        return x == other.x && y == other.y;
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y);
    }

    @Override
    public String toString() {
        return String.format("Position(%d, %d)", x, y);
    }
}

File: src/main/scala/com/gridgame/common/model/Position.scala
package com.gridgame.common.model

import com.gridgame.common.Constants
import java.util.Objects

final class Position(private val x: Int, private val y: Int) {
  if (x < 0 || x >= Constants.GRID_SIZE) {
    throw new IllegalArgumentException(
      s"X coordinate $x out of bounds [0, ${Constants.GRID_SIZE})"
    )
  }
  if (y < 0 || y >= Constants.GRID_SIZE) {
    throw new IllegalArgumentException(
      s"Y coordinate $y out of bounds [0, ${Constants.GRID_SIZE})"
    )
  }

  def getX: Int = x

  def getY: Int = y

  override def equals(obj: Any): Boolean = {
    obj match {
      case that: Position => this.x == that.x && this.y == that.y
      case _ => false
    }
  }

  override def hashCode(): Int = Objects.hash(Int.box(x), Int.box(y))

  override def toString: String = s"Position($x, $y)"
}
